x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/checkpoint/DefaultCheckpointProvider.java

```diff
@@ -311,4 +322,34 @@ public void getCheckpointingInfo(long lastCheckpointNumber, IndexerState nextChe
             getIndexCheckpoints(checkpointsByIndexListener);
         }
     }
+
+    /**
+     * Inspect source changes and report differences
+     *
+     * @param lastSourceIndexes the set of indexes seen in the previous checkpoint
+     * @param newSourceIndexes the set of indexes seen in the new checkpoint
+     */
+    void reportSourceIndexChanges(final Set<String> lastSourceIndexes, final Set<String> newSourceIndexes) {
+        // spam protection: only warn the first time
+        if (newSourceIndexes.isEmpty() && lastSourceIndexes.isEmpty() == false) {
+            String message = "Source did not resolve to any concrete indexes";
+            logger.warn("{} for transform [{}]", message, transformConfig.getId());
+            dataFrameAuditor.warning(transformConfig.getId(), message);
+        } else {
+            Set<String> removedIndexes = Sets.difference(lastSourceIndexes, newSourceIndexes);
+            Set<String> addedIndexes = Sets.difference(newSourceIndexes, lastSourceIndexes);
+
+            if (removedIndexes.size() + addedIndexes.size() > AUDIT_CONCRETED_SOURCE_INDEX_CHANGES) {
+                String message = "Source index resolve found more than " + AUDIT_CONCRETED_SOURCE_INDEX_CHANGES + " changes, ["
+                        + removedIndexes.size() + "] removed indexes, [" + addedIndexes.size() + "] new indexes";
+                logger.debug("{} for transform [{}]", message, transformConfig.getId());
+                dataFrameAuditor.info(transformConfig.getId(), message);
+            } else if (removedIndexes.size() + addedIndexes.size() > 0) {
```

Comment:
I don't understand why the initial `if` clause for `AUDIT_CONCRETED_SOURCE_INDEX_CHANGES` exists if we are still going to log and audit when there are ANY changes?

Reply:
for the 1st case we only report numbers, for the 2nd we log the names

Reply:
missed that :D